test/unittests/t_util.cc

```diff
@@ -215,3 +314,855 @@ TEST(T_Util, IsHttpUrl) {
+  free(buffer);
+}
+
+TEST_F(T_Util, Mutex) {
+  ASSERT_DEATH(LockMutex(static_cast<pthread_mutex_t*>(NULL)), ".*");
+  ASSERT_DEATH(UnlockMutex(static_cast<pthread_mutex_t*>(NULL)), ".*");
+}
+
+TEST_F(T_Util, SwitchCredentials) {
+  // if I am root
+  if (getuid() == 0) {
+    SwitchCredentials(1, 1, true);
+    EXPECT_EQ(1u, geteuid());
+    EXPECT_EQ(1u, getegid());
+    SwitchCredentials(0, 0, false);
+    EXPECT_EQ(0u, geteuid());
+    EXPECT_EQ(0u, getegid());
+  } else {
+    SwitchCredentials(0, 0, false);
+    EXPECT_NE(0u, geteuid());
+    EXPECT_NE(0u, getegid());
+  }
+}
+
+TEST_F(T_Util, FileExists) {
+  string filename = string(sandbox) + "/" + "fileexists.txt";
+  FILE* myfile = fopen(filename.c_str(), "w");
+  fclose(myfile);
+
+  EXPECT_FALSE(FileExists("/my/fake/path/to/file.txt"));
+  EXPECT_TRUE(FileExists(filename));
+}
+
+TEST_F(T_Util, GetFileSize) {
+  string file = CreateFileWithContent("filesize.txt", to_write);
+
+  EXPECT_EQ(-1, GetFileSize("/my/fake/path/to/file.txt"));
+  EXPECT_LT(0, GetFileSize(file));
+}
+
+TEST_F(T_Util, DirectoryExists) {
+  EXPECT_TRUE(DirectoryExists("/usr"));
+  EXPECT_FALSE(DirectoryExists("/fakepath/myfakepath"));
+}
+
+TEST_F(T_Util, SymlinkExists) {
+  string symlinkname = string(sandbox) + "/mysymlink";
+  string filename = CreateFileWithContent("mysymlinkfile.txt", to_write);
+  symlink(filename.c_str(), symlinkname.c_str());
+
+  EXPECT_TRUE(SymlinkExists(symlinkname));
+  EXPECT_FALSE(SymlinkExists("/fakepath/myfakepath"));
+}
+
+TEST_F(T_Util, MkdirDeep) {
+  string mydirectory = string(sandbox) + "/mydirectory";
+  string myfile = CreateFileWithContent("myfile.txt", to_write);
+
+  EXPECT_FALSE(MkdirDeep(string(empty), 0777));
+  EXPECT_TRUE(MkdirDeep(mydirectory, 0777));
+  EXPECT_TRUE(MkdirDeep(string(sandbox) + "/another/another/mydirectory",
+      0777));
+  EXPECT_TRUE(MkdirDeep(mydirectory, 0777));  // already exists, but writable
+  EXPECT_FALSE(MkdirDeep(myfile, 0777));
+}
+
+TEST_F(T_Util, MakeCacheDirectories) {
+  const string path = string(sandbox) + "/cache";
+  MakeCacheDirectories(path, 0777);
+
+  EXPECT_TRUE(DirectoryExists(path + "/txn"));
+  EXPECT_TRUE(DirectoryExists(path + "/quarantaine"));
+  for (int i = 0; i <= 0xff; i++) {
+    char hex[3];
+    snprintf(hex, sizeof(hex), "%02x", i);
+    string current_dir = path + "/" + string(hex);
+    ASSERT_TRUE(DirectoryExists(current_dir));
+  }
+}
+
+TEST_F(T_Util, TryLockFile) {
+  string filename = string(sandbox) + "/trylockfile.txt";
+  int fd;
+
+  EXPECT_EQ(-1, TryLockFile("/fakepath/fakefile.txt"));
+  EXPECT_LE(0, fd = TryLockFile(filename));
+  EXPECT_EQ(-2, TryLockFile(filename));  // second time it fails
+  close(fd);
+}
+
+TEST_F(T_Util, LockFile) {
+  string filename = string(sandbox) + "/lockfile.txt";
+  int retval;
+  int fd;
+
+  EXPECT_EQ(-1, LockFile("/fakepath/fakefile.txt"));
+  EXPECT_LE(0, fd = LockFile(filename));
+
+  tbb::tbb_thread thread(LockFileTest, filename, &retval);
+  sleep(2);
```

Comment:
You don't want to release the main thread when the other is active, but when the other is blocked in the function we are testing here. But if the other is blocked... how is he going to tell the main thread that he actually is? They both have to wait until each other is ready, but at the same time there is no way to warn them.